File: src/slice/converter.rs
//! Converter from Clash config to slice-based format.

use std::collections::HashMap;
use std::net::Ipv4Addr;

use crate::{Result, Target};

use super::SliceWriter;

/// Intermediate representation of a slice.
#[derive(Debug, Clone)]
enum SliceData {
    Domains(Vec<String>, Target),
    CidrV4(Vec<(u32, u8)>, Target),
    CidrV6(Vec<([u8; 16], u8)>, Target),
    GeoIp(Vec<String>, Target),
}

impl SliceData {
    fn target(&self) -> Target {
        match self {
            SliceData::Domains(_, t) => *t,
            SliceData::CidrV4(_, t) => *t,
            SliceData::CidrV6(_, t) => *t,
            SliceData::GeoIp(_, t) => *t,
        }
    }

    fn slice_type(&self) -> SliceType {
        match self {
            SliceData::Domains(_, _) => SliceType::Domain,
            SliceData::CidrV4(_, _) => SliceType::CidrV4,
            SliceData::CidrV6(_, _) => SliceType::CidrV6,
            SliceData::GeoIp(_, _) => SliceType::GeoIp,
        }
    }

    fn can_merge(&self, other: &SliceData) -> bool {
        self.slice_type() == other.slice_type() && self.target() == other.target()
    }

    fn merge(&mut self, other: SliceData) {
        match (self, other) {
            (SliceData::Domains(a, _), SliceData::Domains(b, _)) => a.extend(b),
            (SliceData::CidrV4(a, _), SliceData::CidrV4(b, _)) => a.extend(b),
            (SliceData::CidrV6(a, _), SliceData::CidrV6(b, _)) => a.extend(b),
            (SliceData::GeoIp(a, _), SliceData::GeoIp(b, _)) => a.extend(b),
            _ => {}
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
enum SliceType {
    Domain,
    CidrV4,
    CidrV6,
    GeoIp,
}

/// Clash configuration structure for deserialization.
#[derive(Debug, serde::Deserialize)]
struct ClashConfig {
    #[serde(default)]
    rules: Vec<String>,
    #[serde(rename = "rule-providers", default)]
    rule_providers: HashMap<String, RuleProvider>,
}

/// Clash rule provider configuration.
#[derive(Debug, serde::Deserialize)]
struct RuleProvider {
    behavior: String,
    #[serde(default)]
    rules: Vec<String>,
}

/// Converter from Clash config to slice-based format.
pub struct SliceConverter {
    provider_rules: HashMap<String, Vec<String>>,
}

impl SliceConverter {
    /// Create a new converter.
    pub fn new() -> Self {
        Self {
            provider_rules: HashMap::new(),
        }
    }

    /// Set provider rules directly (for testing or preloaded providers).
    pub fn set_provider_rules(&mut self, name: &str, rules: Vec<String>) {
        self.provider_rules.insert(name.to_string(), rules);
    }

    /// Load provider rules from YAML content.
    pub fn load_provider(&mut self, name: &str, content: &str) -> Result<()> {
        #[derive(serde::Deserialize)]
        struct Payload {
            #[serde(default)]
            payload: Vec<String>,
        }
        let payload: Payload = serde_yaml::from_str(content)?;
        self.provider_rules.insert(name.to_string(), payload.payload);
        Ok(())
    }

    /// Convert Clash config to slice-based binary format.
    ///
    /// Returns the binary data. The fallback target is embedded in the header.
    pub fn convert(&self, yaml_content: &str) -> Result<Vec<u8>> {
        let config: ClashConfig = serde_yaml::from_str(yaml_content)?;

        // Collect slices in rule order
        let mut slices: Vec<SliceData> = Vec::new();
        let mut fallback = Target::Direct;

        for rule_str in &config.rules {
            let parts: Vec<&str> = rule_str.split(',').collect();
            if parts.len() < 2 {
                continue;
            }

            let rule_type = parts[0].trim();
            let target = Target::from_str_lossy(parts.last().unwrap_or(&"PROXY"));

            match rule_type {
                "RULE-SET" => {
                    if parts.len() < 3 {
                        continue;
                    }
                    let provider_name = parts[1].trim();

                    // Get provider behavior from config
                    let behavior = config
                        .rule_providers
                        .get(provider_name)
                        .map(|p| p.behavior.as_str())
                        .unwrap_or("domain");

                    // Get rules from loaded providers or inline
                    let rules = self
                        .provider_rules
                        .get(provider_name)
                        .cloned()
                        .or_else(|| {
                            config
                                .rule_providers
                                .get(provider_name)
                                .map(|p| p.rules.clone())
                        })
                        .unwrap_or_default();

                    self.add_provider_slices(&mut slices, behavior, &rules, target);
                }
                "GEOIP" => {
                    if parts.len() >= 3 {
                        let country = parts[1].trim();
                        if country == "LAN" {
                            add_lan_slices(&mut slices, target);
                        } else {
                            slices.push(SliceData::GeoIp(vec![country.to_string()], target));
                        }
                    }
                }
                "DOMAIN" => {
                    if parts.len() >= 3 {
                        let domain = parts[1].trim();
                        slices.push(SliceData::Domains(vec![domain.to_string()], target));
                    }
                }
                "DOMAIN-SUFFIX" => {
                    if parts.len() >= 3 {
                        let suffix = parts[1].trim();
                        slices.push(SliceData::Domains(vec![suffix.to_string()], target));
                    }
                }
                "IP-CIDR" => {
                    if parts.len() >= 3 {
                        let cidr = parts[1].trim();
                        if let Some((network, prefix_len)) = parse_cidr_v4(cidr) {
                            slices.push(SliceData::CidrV4(vec![(network, prefix_len)], target));
                        }
                    }
                }
                "IP-CIDR6" => {
                    if parts.len() >= 3 {
                        let cidr = parts[1].trim();
                        if let Some((network, prefix_len)) = parse_cidr_v6(cidr) {
                            slices.push(SliceData::CidrV6(vec![(network, prefix_len)], target));
                        }
                    }
                }
                "MATCH" => {
                    fallback = target;
                }
                _ => {}
            }
        }

        // Merge adjacent same-type same-target slices
        let merged_slices = merge_adjacent_slices(slices);

        // Build binary using SliceWriter
        let mut writer = SliceWriter::new(fallback);

        for slice in merged_slices {
            match slice {
                SliceData::Domains(domains, target) => {
                    let domain_refs: Vec<&str> = domains.iter().map(|s| s.as_str()).collect();
                    writer.add_domain_slice(&domain_refs, target)?;
                }
                SliceData::CidrV4(cidrs, target) => {
                    writer.add_cidr_v4_slice(&cidrs, target)?;
                }
                SliceData::CidrV6(cidrs, target) => {
                    writer.add_cidr_v6_slice(&cidrs, target)?;
                }
                SliceData::GeoIp(countries, target) => {
                    let country_refs: Vec<&str> = countries.iter().map(|s| s.as_str()).collect();
                    writer.add_geoip_slice(&country_refs, target)?;
                }
            }
        }

        writer.build()
    }

    fn add_provider_slices(
        &self,
        slices: &mut Vec<SliceData>,
        behavior: &str,
        rules: &[String],
        target: Target,
    ) {
        match behavior {
            "domain" => {
                let mut domains = Vec::new();
                for rule in rules {
                    let rule = rule.trim();
                    if rule.is_empty() || rule.starts_with('#') {
                        continue;
                    }

                    // Convert Clash suffix format to our format
                    let domain = if rule.starts_with("+.") {
                        rule[2..].to_string()
                    } else if rule.starts_with('.') {
                        rule[1..].to_string()
                    } else {
                        rule.to_string()
                    };

                    domains.push(domain);
                }
                if !domains.is_empty() {
                    slices.push(SliceData::Domains(domains, target));
                }
            }
            "ipcidr" => {
                let mut v4_cidrs = Vec::new();
                let mut v6_cidrs = Vec::new();

                for rule in rules {
                    let rule = rule.trim();
                    if rule.is_empty() || rule.starts_with('#') {
                        continue;
                    }

                    if let Some((network, prefix_len)) = parse_cidr_v4(rule) {
                        v4_cidrs.push((network, prefix_len));
                    } else if let Some((network, prefix_len)) = parse_cidr_v6(rule) {
                        v6_cidrs.push((network, prefix_len));
                    }
                }

                if !v4_cidrs.is_empty() {
                    slices.push(SliceData::CidrV4(v4_cidrs, target));
                }
                if !v6_cidrs.is_empty() {
                    slices.push(SliceData::CidrV6(v6_cidrs, target));
                }
            }
            "classical" => {
                for rule in rules {
                    self.parse_classical_rule(slices, rule, target);
                }
            }
            _ => {}
        }
    }

    fn parse_classical_rule(&self, slices: &mut Vec<SliceData>, rule: &str, target: Target) {
        let parts: Vec<&str> = rule.split(',').collect();
        if parts.len() < 2 {
            return;
        }

        let rule_type = parts[0].trim();
        let value = parts[1].trim();

        match rule_type {
            "DOMAIN" | "DOMAIN-SUFFIX" => {
                slices.push(SliceData::Domains(vec![value.to_string()], target));
            }
            "IP-CIDR" => {
                if let Some((network, prefix_len)) = parse_cidr_v4(value) {
                    slices.push(SliceData::CidrV4(vec![(network, prefix_len)], target));
                }
            }
            "IP-CIDR6" => {
                if let Some((network, prefix_len)) = parse_cidr_v6(value) {
                    slices.push(SliceData::CidrV6(vec![(network, prefix_len)], target));
                }
            }
            "GEOIP" => {
                if value == "LAN" {
                    add_lan_slices(slices, target);
                } else {
                    slices.push(SliceData::GeoIp(vec![value.to_string()], target));
                }
            }
            _ => {}
        }
    }
}

impl Default for SliceConverter {
    fn default() -> Self {
        Self::new()
    }
}

/// Merge adjacent slices with same type and target.
fn merge_adjacent_slices(slices: Vec<SliceData>) -> Vec<SliceData> {
    if slices.is_empty() {
        return slices;
    }

    let mut result: Vec<SliceData> = Vec::new();

    for slice in slices {
        if let Some(last) = result.last_mut() {
            if last.can_merge(&slice) {
                last.merge(slice);
                continue;
            }
        }
        result.push(slice);
    }

    result
}

/// Add LAN CIDR slices.
fn add_lan_slices(slices: &mut Vec<SliceData>, target: Target) {
    // IPv4 private ranges
    slices.push(SliceData::CidrV4(
        vec![
            (0x7F000000, 8),  // 127.0.0.0/8
            (0x0A000000, 8),  // 10.0.0.0/8
            (0xAC100000, 12), // 172.16.0.0/12
            (0xC0A80000, 16), // 192.168.0.0/16
            (0xA9FE0000, 16), // 169.254.0.0/16
        ],
        target,
    ));

    // IPv6 private ranges
    let fc00: [u8; 16] = [0xFC, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0];
    let fe80: [u8; 16] = [0xFE, 0x80, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0];
    let localhost: [u8; 16] = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1];

    slices.push(SliceData::CidrV6(
        vec![(fc00, 7), (fe80, 10), (localhost, 128)],
        target,
    ));
}

/// Parse IPv4 CIDR string.
fn parse_cidr_v4(cidr: &str) -> Option<(u32, u8)> {
    let parts: Vec<&str> = cidr.split('/').collect();
    if parts.len() != 2 {
        return None;
    }

    let addr: Ipv4Addr = parts[0].parse().ok()?;
    let prefix_len: u8 = parts[1].parse().ok()?;

    Some((u32::from(addr), prefix_len))
}

/// Parse IPv6 CIDR string.
fn parse_cidr_v6(cidr: &str) -> Option<([u8; 16], u8)> {
    let parts: Vec<&str> = cidr.split('/').collect();
    if parts.len() != 2 {
        return None;
    }

    let addr: std::net::Ipv6Addr = parts[0].parse().ok()?;
    let prefix_len: u8 = parts[1].parse().ok()?;

    Some((addr.octets(), prefix_len))
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::slice::SliceReader;

    #[test]
    fn test_simple_rules() {
        let yaml = r#"
rules:
  - DOMAIN,google.com,PROXY
  - DOMAIN-SUFFIX,youtube.com,PROXY
  - GEOIP,CN,DIRECT
  - MATCH,PROXY
"#;

        let converter = SliceConverter::new();
        let data = converter.convert(yaml).unwrap();

        let reader = SliceReader::from_bytes(&data).unwrap();
        assert_eq!(reader.fallback(), Target::Proxy);
        assert_eq!(reader.match_domain("google.com"), Some(Target::Proxy));
        assert_eq!(reader.match_domain("youtube.com"), Some(Target::Proxy));
        assert_eq!(reader.match_domain("www.youtube.com"), Some(Target::Proxy));
        assert_eq!(reader.match_geoip("CN"), Some(Target::Direct));
    }

    #[test]
    fn test_rule_providers() {
        let yaml = r#"
rule-providers:
  proxy-domains:
    type: http
    behavior: domain
    rules:
      - google.com
      - "+.googleapis.com"

rules:
  - RULE-SET,proxy-domains,PROXY
  - MATCH,DIRECT
"#;

        let converter = SliceConverter::new();
        let data = converter.convert(yaml).unwrap();

        let reader = SliceReader::from_bytes(&data).unwrap();
        assert_eq!(reader.fallback(), Target::Direct);
        assert_eq!(reader.match_domain("google.com"), Some(Target::Proxy));
        assert_eq!(
            reader.match_domain("www.googleapis.com"),
            Some(Target::Proxy)
        );
    }

    #[test]
    fn test_merge_adjacent_slices() {
        let yaml = r#"
rules:
  - DOMAIN,a.com,PROXY
  - DOMAIN,b.com,PROXY
  - DOMAIN,c.com,DIRECT
  - DOMAIN,d.com,PROXY
  - MATCH,DIRECT
"#;

        let converter = SliceConverter::new();
        let data = converter.convert(yaml).unwrap();

        let reader = SliceReader::from_bytes(&data).unwrap();
        // a.com and b.com should be merged (same type, same target)
        // c.com is separate (different target)
        // d.com is separate (not adjacent to b.com after c.com)
        assert_eq!(reader.slice_count(), 3);

        assert_eq!(reader.match_domain("a.com"), Some(Target::Proxy));
        assert_eq!(reader.match_domain("b.com"), Some(Target::Proxy));
        assert_eq!(reader.match_domain("c.com"), Some(Target::Direct));
        assert_eq!(reader.match_domain("d.com"), Some(Target::Proxy));
    }

    #[test]
    fn test_ordering_preserved() {
        // cn.bing.com -> Direct first, then bing.com -> Proxy
        let yaml = r#"
rules:
  - DOMAIN-SUFFIX,cn.bing.com,DIRECT
  - DOMAIN-SUFFIX,bing.com,PROXY
  - MATCH,DIRECT
"#;

        let converter = SliceConverter::new();
        let data = converter.convert(yaml).unwrap();

        let reader = SliceReader::from_bytes(&data).unwrap();

        // cn.bing.com should match first slice (Direct)
        assert_eq!(reader.match_domain("cn.bing.com"), Some(Target::Direct));
        assert_eq!(
            reader.match_domain("www.cn.bing.com"),
            Some(Target::Direct)
        );

        // bing.com should match second slice (Proxy)
        assert_eq!(reader.match_domain("bing.com"), Some(Target::Proxy));
        assert_eq!(reader.match_domain("www.bing.com"), Some(Target::Proxy));
    }

    #[test]
    fn test_cidr_rules() {
        let yaml = r#"
rules:
  - IP-CIDR,10.0.0.0/8,DIRECT
  - IP-CIDR6,fc00::/7,DIRECT
  - MATCH,PROXY
"#;

        let converter = SliceConverter::new();
        let data = converter.convert(yaml).unwrap();

        let reader = SliceReader::from_bytes(&data).unwrap();

        assert_eq!(
            reader.match_ip("10.1.2.3".parse().unwrap()),
            Some(Target::Direct)
        );
        assert_eq!(
            reader.match_ip("fc00::1".parse().unwrap()),
            Some(Target::Direct)
        );
        assert_eq!(reader.match_ip("8.8.8.8".parse().unwrap()), None);
    }

    #[test]
    fn test_lan_geoip() {
        let yaml = r#"
rules:
  - GEOIP,LAN,DIRECT
  - MATCH,PROXY
"#;

        let converter = SliceConverter::new();
        let data = converter.convert(yaml).unwrap();

        let reader = SliceReader::from_bytes(&data).unwrap();

        // LAN IPs should match
        assert_eq!(
            reader.match_ip("10.0.0.1".parse().unwrap()),
            Some(Target::Direct)
        );
        assert_eq!(
            reader.match_ip("192.168.1.1".parse().unwrap()),
            Some(Target::Direct)
        );
        assert_eq!(
            reader.match_ip("fc00::1".parse().unwrap()),
            Some(Target::Direct)
        );
    }

    #[test]
    fn test_external_provider() {
        let yaml = r#"
rule-providers:
  external:
    type: http
    behavior: domain
    url: https://example.com/rules.yaml

rules:
  - RULE-SET,external,PROXY
  - MATCH,DIRECT
"#;

        let mut converter = SliceConverter::new();
        converter.set_provider_rules(
            "external",
            vec!["google.com".to_string(), "youtube.com".to_string()],
        );

        let data = converter.convert(yaml).unwrap();

        let reader = SliceReader::from_bytes(&data).unwrap();
        assert_eq!(reader.match_domain("google.com"), Some(Target::Proxy));
        assert_eq!(reader.match_domain("youtube.com"), Some(Target::Proxy));
    }

    #[test]
    fn test_ipcidr_provider() {
        let yaml = r#"
rule-providers:
  private:
    type: http
    behavior: ipcidr
    rules:
      - 10.0.0.0/8
      - 172.16.0.0/12

rules:
  - RULE-SET,private,DIRECT
  - MATCH,PROXY
"#;

        let converter = SliceConverter::new();
        let data = converter.convert(yaml).unwrap();

        let reader = SliceReader::from_bytes(&data).unwrap();
        assert_eq!(
            reader.match_ip("10.1.2.3".parse().unwrap()),
            Some(Target::Direct)
        );
        assert_eq!(
            reader.match_ip("172.20.0.1".parse().unwrap()),
            Some(Target::Direct)
        );
    }
}

File: internal/clash/converter.go
// Package clash provides conversion from Clash YAML configurations to K2RULEV2 binary format.
package clash

import (
	"encoding/binary"
	"net"
	"strings"

	"gopkg.in/yaml.v3"

	"github.com/kaitu-io/k2rule/internal/slice"
)

// target constants matching Rust and Go codebase values.
const (
	targetDirect uint8 = 0
	targetProxy  uint8 = 1
	targetReject uint8 = 2
)

// parseTarget parses a Clash target string to its uint8 value.
// Unknown values default to Proxy.
func parseTarget(s string) uint8 {
	switch strings.ToUpper(strings.TrimSpace(s)) {
	case "DIRECT":
		return targetDirect
	case "REJECT":
		return targetReject
	default:
		return targetProxy
	}
}

// sliceKind identifies the type of a collected slice.
type sliceKind int

const (
	kindDomain sliceKind = iota
	kindCidrV4
	kindCidrV6
	kindGeoIP
)

// sliceData holds the intermediate representation of a slice before writing.
type sliceData struct {
	kind    sliceKind
	target  uint8
	domains []string
	cidrV4  [][2]uint32 // [network_u32, prefix_len]
	cidrV6N [][16]byte
	cidrV6P []uint8
	geoips  []string
}

func (sd *sliceData) canMerge(other *sliceData) bool {
	return sd.kind == other.kind && sd.target == other.target
}

func (sd *sliceData) merge(other *sliceData) {
	switch sd.kind {
	case kindDomain:
		sd.domains = append(sd.domains, other.domains...)
	case kindCidrV4:
		sd.cidrV4 = append(sd.cidrV4, other.cidrV4...)
	case kindCidrV6:
		sd.cidrV6N = append(sd.cidrV6N, other.cidrV6N...)
		sd.cidrV6P = append(sd.cidrV6P, other.cidrV6P...)
	case kindGeoIP:
		sd.geoips = append(sd.geoips, other.geoips...)
	}
}

// clashConfig represents the top-level Clash configuration YAML structure.
type clashConfig struct {
	Rules         []string                `yaml:"rules"`
	RuleProviders map[string]ruleProvider `yaml:"rule-providers"`
}

// ruleProvider represents a Clash rule provider configuration.
type ruleProvider struct {
	Behavior string   `yaml:"behavior"`
	Rules    []string `yaml:"rules"`
	URL      string   `yaml:"url"`
}

// SliceConverter converts Clash YAML configurations to K2RULEV2 binary format.
type SliceConverter struct {
	// providerRules holds preloaded or externally set provider rules.
	// Keys are provider names; values are lists of rule strings.
	providerRules map[string][]string
}

// NewSliceConverter creates a new SliceConverter.
func NewSliceConverter() *SliceConverter {
	return &SliceConverter{
		providerRules: make(map[string][]string),
	}
}

// SetProviderRules sets provider rules directly (for testing or preloaded providers).
// This overrides any rules loaded via LoadProvider for the same name.
func (c *SliceConverter) SetProviderRules(name string, rules []string) {
	c.providerRules[name] = rules
}

// LoadProvider loads provider rules from content string.
// Supports both YAML "payload:" format and plain text (one rule per line).
func (c *SliceConverter) LoadProvider(name, content string) error {
	// Try YAML payload format first
	type payloadDoc struct {
		Payload []string `yaml:"payload"`
	}
	var doc payloadDoc
	if err := yaml.Unmarshal([]byte(content), &doc); err == nil && len(doc.Payload) > 0 {
		c.providerRules[name] = doc.Payload
		return nil
	}

	// Fall back to plain text format (one entry per line)
	var rules []string
	for _, line := range strings.Split(content, "\n") {
		line = strings.TrimSpace(line)
		if line == "" || strings.HasPrefix(line, "#") {
			continue
		}
		rules = append(rules, line)
	}
	c.providerRules[name] = rules
	return nil
}

// Convert parses a Clash YAML configuration and produces K2RULEV2 binary data.
func (c *SliceConverter) Convert(yamlContent string) ([]byte, error) {
	var config clashConfig
	if err := yaml.Unmarshal([]byte(yamlContent), &config); err != nil {
		return nil, err
	}

	var slices []*sliceData
	fallback := targetDirect

	for _, ruleStr := range config.Rules {
		parts := strings.Split(ruleStr, ",")
		if len(parts) < 2 {
			continue
		}

		ruleType := strings.TrimSpace(parts[0])
		targetStr := strings.TrimSpace(parts[len(parts)-1])
		target := parseTarget(targetStr)

		switch ruleType {
		case "DOMAIN":
			if len(parts) < 3 {
				continue
			}
			domain := strings.TrimSpace(parts[1])
			slices = appendOrMerge(slices, &sliceData{
				kind:    kindDomain,
				target:  target,
				domains: []string{domain},
			})

		case "DOMAIN-SUFFIX":
			if len(parts) < 3 {
				continue
			}
			suffix := strings.TrimSpace(parts[1])
			slices = appendOrMerge(slices, &sliceData{
				kind:    kindDomain,
				target:  target,
				domains: []string{suffix},
			})

		case "IP-CIDR":
			if len(parts) < 3 {
				continue
			}
			cidr := strings.TrimSpace(parts[1])
			network, prefixLen, ok := parseCIDRv4(cidr)
			if !ok {
				continue
			}
			slices = appendOrMerge(slices, &sliceData{
				kind:   kindCidrV4,
				target: target,
				cidrV4: [][2]uint32{{network, uint32(prefixLen)}},
			})

		case "IP-CIDR6":
			if len(parts) < 3 {
				continue
			}
			cidr := strings.TrimSpace(parts[1])
			network, prefixLen, ok := parseCIDRv6(cidr)
			if !ok {
				continue
			}
			slices = appendOrMerge(slices, &sliceData{
				kind:    kindCidrV6,
				target:  target,
				cidrV6N: [][16]byte{network},
				cidrV6P: []uint8{prefixLen},
			})

		case "GEOIP":
			if len(parts) < 3 {
				continue
			}
			country := strings.TrimSpace(parts[1])
			if strings.ToUpper(country) == "LAN" {
				addLANSlices(&slices, target)
			} else {
				slices = appendOrMerge(slices, &sliceData{
					kind:   kindGeoIP,
					target: target,
					geoips: []string{country},
				})
			}

		case "RULE-SET":
			if len(parts) < 3 {
				continue
			}
			providerName := strings.TrimSpace(parts[1])

			// Get behavior from config
			behavior := "domain"
			if p, ok := config.RuleProviders[providerName]; ok {
				behavior = p.Behavior
			}

			// Get rules: prefer externally loaded rules, then inline rules from config
			rules, ok := c.providerRules[providerName]
			if !ok {
				if p, ok2 := config.RuleProviders[providerName]; ok2 {
					rules = p.Rules
				}
			}

			c.addProviderSlices(&slices, behavior, rules, target)

		case "MATCH":
			fallback = target
		}
	}

	// Build binary using SliceWriter
	writer := slice.NewSliceWriter(fallback)
	for _, sd := range slices {
		if err := writeSliceData(writer, sd); err != nil {
			return nil, err
		}
	}
	return writer.Build()
}

// appendOrMerge appends a new sliceData to the list, merging with the last
// entry if they have the same kind and target.
func appendOrMerge(slices []*sliceData, sd *sliceData) []*sliceData {
	if len(slices) > 0 {
		last := slices[len(slices)-1]
		if last.canMerge(sd) {
			last.merge(sd)
			return slices
		}
	}
	return append(slices, sd)
}

// addProviderSlices expands a RULE-SET provider into sliceData entries.
func (c *SliceConverter) addProviderSlices(slices *[]*sliceData, behavior string, rules []string, target uint8) {
	switch strings.ToLower(behavior) {
	case "domain":
		var domains []string
		for _, rule := range rules {
			rule = strings.TrimSpace(rule)
			if rule == "" || strings.HasPrefix(rule, "#") {
				continue
			}
			// Convert Clash suffix format (+. or leading .) to plain domain
			if strings.HasPrefix(rule, "+.") {
				rule = rule[2:]
			} else if strings.HasPrefix(rule, ".") {
				rule = rule[1:]
			}
			domains = append(domains, rule)
		}
		if len(domains) > 0 {
			*slices = appendOrMerge(*slices, &sliceData{
				kind:    kindDomain,
				target:  target,
				domains: domains,
			})
		}

	case "ipcidr":
		var v4 [][2]uint32
		var v6N [][16]byte
		var v6P []uint8
		for _, rule := range rules {
			rule = strings.TrimSpace(rule)
			if rule == "" || strings.HasPrefix(rule, "#") {
				continue
			}
			if network, pfl, ok := parseCIDRv4(rule); ok {
				v4 = append(v4, [2]uint32{network, uint32(pfl)})
			} else if network6, pfl6, ok6 := parseCIDRv6(rule); ok6 {
				v6N = append(v6N, network6)
				v6P = append(v6P, pfl6)
			}
		}
		if len(v4) > 0 {
			*slices = appendOrMerge(*slices, &sliceData{
				kind:   kindCidrV4,
				target: target,
				cidrV4: v4,
			})
		}
		if len(v6N) > 0 {
			*slices = appendOrMerge(*slices, &sliceData{
				kind:    kindCidrV6,
				target:  target,
				cidrV6N: v6N,
				cidrV6P: v6P,
			})
		}

	case "classical":
		for _, rule := range rules {
			rule = strings.TrimSpace(rule)
			if rule == "" || strings.HasPrefix(rule, "#") {
				continue
			}
			c.parseClassicalRule(slices, rule, target)
		}
	}
}

// parseClassicalRule parses a single classical-format rule line and appends to slices.
func (c *SliceConverter) parseClassicalRule(slices *[]*sliceData, rule string, target uint8) {
	parts := strings.Split(rule, ",")
	if len(parts) < 2 {
		return
	}

	ruleType := strings.TrimSpace(parts[0])
	value := strings.TrimSpace(parts[1])

	switch ruleType {
	case "DOMAIN", "DOMAIN-SUFFIX":
		*slices = appendOrMerge(*slices, &sliceData{
			kind:    kindDomain,
			target:  target,
			domains: []string{value},
		})
	case "IP-CIDR":
		if network, pfl, ok := parseCIDRv4(value); ok {
			*slices = appendOrMerge(*slices, &sliceData{
				kind:   kindCidrV4,
				target: target,
				cidrV4: [][2]uint32{{network, uint32(pfl)}},
			})
		}
	case "IP-CIDR6":
		if network6, pfl6, ok := parseCIDRv6(value); ok {
			*slices = appendOrMerge(*slices, &sliceData{
				kind:    kindCidrV6,
				target:  target,
				cidrV6N: [][16]byte{network6},
				cidrV6P: []uint8{pfl6},
			})
		}
	case "GEOIP":
		if strings.ToUpper(value) == "LAN" {
			addLANSlices(slices, target)
		} else {
			*slices = appendOrMerge(*slices, &sliceData{
				kind:   kindGeoIP,
				target: target,
				geoips: []string{value},
			})
		}
	}
}

// addLANSlices expands the LAN GeoIP shortcut into IPv4 and IPv6 private ranges.
func addLANSlices(slices *[]*sliceData, target uint8) {
	// IPv4 private ranges: 127.0.0.0/8, 10.0.0.0/8, 172.16.0.0/12, 192.168.0.0/16, 169.254.0.0/16
	*slices = appendOrMerge(*slices, &sliceData{
		kind:   kindCidrV4,
		target: target,
		cidrV4: [][2]uint32{
			{0x7F000000, 8},  // 127.0.0.0/8
			{0x0A000000, 8},  // 10.0.0.0/8
			{0xAC100000, 12}, // 172.16.0.0/12
			{0xC0A80000, 16}, // 192.168.0.0/16
			{0xA9FE0000, 16}, // 169.254.0.0/16
		},
	})

	// IPv6 private ranges: fc00::/7, fe80::/10, ::1/128
	fc00 := [16]byte{0xFC, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0}
	fe80 := [16]byte{0xFE, 0x80, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0}
	loop := [16]byte{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1}

	*slices = appendOrMerge(*slices, &sliceData{
		kind:    kindCidrV6,
		target:  target,
		cidrV6N: [][16]byte{fc00, fe80, loop},
		cidrV6P: []uint8{7, 10, 128},
	})
}

// writeSliceData writes a sliceData entry to the SliceWriter.
func writeSliceData(w *slice.SliceWriter, sd *sliceData) error {
	switch sd.kind {
	case kindDomain:
		return w.AddDomainSlice(sd.domains, sd.target)
	case kindCidrV4:
		return w.AddCidrV4Slice(sd.cidrV4, sd.target)
	case kindCidrV6:
		return w.AddCidrV6SliceRaw(sd.cidrV6N, sd.cidrV6P, sd.target)
	case kindGeoIP:
		return w.AddGeoIPSlice(sd.geoips, sd.target)
	}
	return nil
}

// parseCIDRv4 parses an IPv4 CIDR string and returns (network_u32, prefix_len, ok).
func parseCIDRv4(cidr string) (uint32, uint8, bool) {
	ip, ipnet, err := net.ParseCIDR(cidr)
	if err != nil {
		return 0, 0, false
	}
	ip4 := ip.To4()
	if ip4 == nil {
		// Try network address
		ip4 = ipnet.IP.To4()
	}
	if ip4 == nil {
		return 0, 0, false
	}
	ones, bits := ipnet.Mask.Size()
	if bits != 32 {
		return 0, 0, false
	}
	network := binary.BigEndian.Uint32(ip4)
	// Mask to network address
	network = network & (^uint32(0) << uint(32-ones))
	return network, uint8(ones), true
}

// parseCIDRv6 parses an IPv6 CIDR string and returns (network_bytes, prefix_len, ok).
func parseCIDRv6(cidr string) ([16]byte, uint8, bool) {
	_, ipnet, err := net.ParseCIDR(cidr)
	if err != nil {
		return [16]byte{}, 0, false
	}
	ip16 := ipnet.IP.To16()
	if ip16 == nil {
		return [16]byte{}, 0, false
	}
	ones, bits := ipnet.Mask.Size()
	if bits != 128 {
		return [16]byte{}, 0, false
	}
	var network [16]byte
	copy(network[:], ip16)
	return network, uint8(ones), true
}
